graphql_compiler/schema_transformation/rename_query.py

```diff
@@ -0,0 +1,96 @@
+# Copyright 2019-present Kensho Technologies, LLC.
+from copy import deepcopy
+
+from graphql.language import ast as ast_types
+from graphql.language.visitor import Visitor, visit
+
+from ..exceptions import GraphQLValidationError
+
+
+def rename_query(ast, renamings):
```

Comment:
Any reason not to take a `RenamedSchemaDescriptor` object? Then we have a somewhat stronger belief that the renamings were actually computed via a call to `rename_schema` and are therefore valid.

Reply:
That's a good point. I started by thinking that rename_query would take in a `RenamedSchemaDescriptor`, but moved away from that idea for several reasons, but there are definitely arguments on both sides.

Taking in a `RenamedSchemaDescriptor` means that rename_query is probably being used in conjunction with rename_schema as expected, and it also allows validity checking (if necessary) since we have access to its schema.

One issue though is that I think users may be confused about which direction rename_query operates, if the input is a `RenamedSchemaDescriptor`. Suppose the user previous renamed schema S to S'. The intended use of rename_query will change query Q' written in terms of types in S' into a query Q written in terms of types in S. However, a user may think that rename_query instead changes Q to Q', in the same direction as rename_schema. Passing in a renamings dictionary doesn't cause this kind of confusion.

Reply:
On the other hand, we only want someone to use the function the reversed_named map of `RenamedSchemaDescriptor`, so it is useful to establish this dependency.

Reply:
However, we'll make this dependency clear when we add a section to the README.md about schema merging so I slightly prefer keeping it the way it is.

Reply:
Here's what I'm concerned about. I believe that `renamings` here is actually a reverse name map – the opposite of the `renamings` used for schema renaming. I suspect _this_ is more likely to cause the directionality confusion @qqi0O0 mentioned, compared to taking in a `RenamedSchemaDescriptor`.

Option 1 is to make it impossible to misuse the code, by requiring a `RenamedSchemaDescriptor` object to be used, which essentially proves that the user called the right functions in the right order. Option 2 is to write instructions into the `README.md` and pray that people read it (which I suspect they won't – it's huge!).

I'm quite certain I prefer option 1 here.

Reply:
It's true that the `renamings` here is the reverse of the `renamings` used in schema renaming, for the use case that we have in mind. Nonetheless, if we pass in a dict here, it seems clear what things will get renamed to what -- the keys in `renamings` will get renamed to their corresponding values.
On the other hand, if we pass in a `RenamedSchemaDescriptor` here, it's not clear what will be renamed to what and what mapping of names will be used -- whether it's the same `renamings` dict that produced the renamed schema, or the `reverse_name_map` in the `RenamedSchemaDescriptor`. For that reason I think passing in a `RenamedSchemaDescriptor` is more potentially misleading.